Add runtime assert

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98878

This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints and intermediate constraints into runtime assertions utilizing `aten._assert_async.msg`.

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2. Explore adding non-async version of assert. 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @soumith @desertfire

tugsbayasgalan committed:
Add runtime assert

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
                 r.node.meta["val"] = self.current_node.meta["val"]
             return r
 
+    class AddRuntimeAssertsInInputConstraint(torch.fx.interpreter.Transformer):
```
This seems complicated and I am not entirely sure what it does and I don't see it described in PR desc

tugsbayasgalan replied:
Just updated the PR description. It basically convert input constraints into runtime asserts per specified dimension.

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -245,12 +245,25 @@ def inner(self: "InstructionTranslatorBase", inst: Instruction):
                 return
 
-            # Manually insert torch._assert instead of python assert and jump over
+            # Manually insert torch._assert_async instead of python assert and jump over
```
One interesting question is that, even if we capture the assert, should we also transform it async? If the original assert was an `assert XXX`, the user might be expecting a DtoH sync to happen, and maybe we should respect that, instead of silently turning it into async asserts that can trash the cuda context.

gmagogsfm replied:
Yeah that seems like a problem, but we don't have an `assert` that is blocking, maybe we should add one instead of using assert_async

tugsbayasgalan replied:
I see. I am not very familiar with CUDA so it will take me sometime to get it right :). I can do it in the follow up diff.

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -245,12 +245,25 @@ def inner(self: "InstructionTranslatorBase", inst: Instruction):
+                "call_function",
+                torch.scalar_tensor,
+                *proxy_args_kwargs((value,), {})
+            )
```
Hmm, what if value is a TensorVariable?

tugsbayasgalan replied:
Good catch! If tensor, we just pass it into assert directly

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -813,6 +886,10 @@ def graph_with_interpreter(*args):
 
+    new_graph = AddRuntimeAssertsInInputConstraint(
+        new_graph,
+    ).transform()
```
Hmmph. At some point we should probably start putting these in their own module. It's also annoying that these are inner class definitions as you cannot easily tell if they closed over things or not.

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
                 r.node.meta["val"] = self.current_node.meta["val"]
             return r
 
+    class AddRuntimeAssertsInInputConstraint(torch.fx.interpreter.Transformer):
```
Hmm, I am still having some high level conceptual problems with this.

1. Does it really need to be fused into this function? This seems like something that would be perfectly happy to live as its own, separate transformation pass. You don't have to insert these runtime asserts, right? Maybe the input constraints are being checked some other way.
2. Are you sure these constraints are even what you want? The value ranges will typically look like [2, inf], but remember that 0/1 specialization says that we are unsoundly hoping the graph also works for size 1. You don't have any tests printing the generated asserts so I can't tell if you are generating an assert that's too strict.
3. The asserts, as generated, are very easy to accidentally DCE. Are we producing a functionalized graph or not here?

tugsbayasgalan replied:
1) We don't need to fuse it with this function. I just put it here as there was no other place to put it :) 
2) I think once the @avikchaudhuri 's work on surfacing guards and having more constrained input constraints, they can be used as asserts in the runtime. For now, they are just asserting the value is from 2 to inf. I like runtime asserts as they are very explicit about what we are protecting on. 
3) This PR doesn't produce functionalized graph. But, I am experimenting with the next PR to use AOTAutograd functionalization. So far, I was able to get following graph:
```
# source code
class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("buffer1", torch.ones(6, 1))
        self.register_buffer("buffer2", torch.ones(6, 2))

    def forward(self, x, y):
        self.buffer1.add_(2)
        x.add_(3)
        assert x.shape[0] > 4
        return x.sum() + self.buffer1.sum()


# resulting graph
def forward(self, arg1_1, arg2_1):
    _tensor_constant0 = self._tensor_constant0
    add = torch.ops.aten.add.Tensor(_tensor_constant0, 2);  _tensor_constant0 = None
    add_1 = torch.ops.aten.add.Tensor(arg1_1, 3)
    sym_size = torch.ops.aten.sym_size(arg1_1, 0);  arg1_1 = None
    gt = sym_size > 4;  sym_size = None
    scalar_tensor = torch.ops.aten.scalar_tensor.default(gt, device = device(type='cpu'), pin_memory = False);  gt = None
    _assert_async = torch.ops.aten._assert_async.msg(scalar_tensor, 'assertion error');  scalar_tensor = None
    sum_1 = torch.ops.aten.sum.default(add_1)
    sum_2 = torch.ops.aten.sum.default(add)
    add_2 = torch.ops.aten.add.Tensor(sum_1, sum_2);  sum_1 = sum_2 = None
    return (add, add_1, add_2)
 ```
Note that here we are asserting on the old value of input argument which can be unsound. So i was thinking about just replacing the returned output in the assert statement directly.

tugsbayasgalan replied:
Actually this might be problematic. It doesn't seem that easy to just reuse the output directly in the assert because we would then need to keep track whether assert comes before the inplace mutation or not.

avikchaudhuri replied:
Just so I'm following, the above code is incorrect because `sym_size` should depend on `add_1` instead of `arg1_1`, right?

tugsbayasgalan replied:
yeah well it is kind of correct in the sense that assert will still work. For shape metadata mutation cases, it seems to capture the correct shape to assert. I am trying to verify this hypothesis now.

ezyang replied:
> For now, they are just asserting the value is from 2 to inf. I like runtime asserts as they are very explicit about what we are protecting on.

What I am saying is, you probably don't want a 2 to inf bound. Because the agreement we had when we decided to do unsound 0/1 spec because we were going to guess that the graph would work anyway with size 1. If you burn in the assert here, then the model will definitely not work with size 1!

tugsbayasgalan replied:
Ah yeah good point! These asserts are not very useful until we have an API that users can specify a range

avikchaudhuri commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
+            super().__init__(m)
+            self.count = 0
+            self.constraints_id_to_constraint = defaultdict(list)
+            if constraints is not None:
```
Is this captured from above? If so then why bother making self.constraints_id_to_constraint, maybe compute it above too...

tugsbayasgalan replied:
Yeah I can :)

tugsbayasgalan replied:
Actually left it here because this map is only needed within this pass.

avikchaudhuri commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
+                min_int_val = _convert_to_int(constraint_range.vr.lower)
+                max_int_val = _convert_to_int(constraint_range.vr.upper)
+
+                if min_int_val is None and max_int_val is None:
```
I don't think these can be None? You should be able to trust upstream and assert in _convert_to_int I think...

tugsbayasgalan replied:
It will be None if the expression is not something we can convert to integer (complicated guard expression).

avikchaudhuri commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
+                dim = self.tracer.create_proxy('call_function', torch.ops.aten.sym_size, (arg, constraint.dim), {})
+                assert_msg = f"Input #{self.count}'s dimension #{constraint.dim} size is outside of supported dynamic range"
+
+                if min_int_val:
```
Why this check?

tugsbayasgalan replied:
If the min_val is somehow None, we want skip asserting.

avikchaudhuri commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
+                    tensor_gt = self.tracer.create_proxy('call_function', torch.scalar_tensor, (gt,), {})
+                    self.tracer.create_proxy('call_function', torch.ops.aten._assert_async.msg, (tensor_gt, assert_msg), {})
+
+                if max_int_val:
```
Why this check?

tugsbayasgalan replied:
If the max_val is somehow None (when expression is too complicated to make it concrete integer), we want skip asserting.

avikchaudhuri commented on torch/_dynamo/eval_frame.py:
```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
+                # TODO ignore expressions for now
+                return None
+
+            for constraint in constraints:
```
It doesn't seem like a good idea to do each of these transformations by hand...rather, can't we generate Python code and trace it in? In the future there may be more kinds of constraints...

tugsbayasgalan replied:
Yep I can explore this option in the follow up diff.

avikchaudhuri commented on torch/_dynamo/eval_frame.py:
```diff
@@ -813,6 +886,10 @@ def graph_with_interpreter(*args):
         graph,
     ).transform()
 
+    new_graph = AddRuntimeAssertsInInputConstraint(
```
This is not the full set of runtime assertions. Anything we specialized on technically should be checked as well.

tugsbayasgalan replied:
Yeah this is only for Input constraints. I will make similar change for inline constraints as well

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -790,41 +963,11 @@ def result_capturing_wrapper(*graph_inputs):
     flat_both = list(graph_captured_result) + flat_args
     matched_output_elements_positions = produce_matching(flat_both, flat_results_traced)
 
-    class ChangeInputOutputSignature(torch.fx.interpreter.Transformer):
```
No change here right

tugsbayasgalan replied:
Nope i just moved it out of the export call

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -790,41 +963,11 @@ def result_capturing_wrapper(*graph_inputs):
+        graph,
+        constraints,
+        flat_args,
+    ).transform()
```
So... you have moved the class out (good!) but you are still unconditionally calling the ops here?

tugsbayasgalan replied:
What do you mean unconditionally calling the ops here? Do you mean we want to do this transform under a flag?

ezyang replied:
This is subsumed by the convo below

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
         )
 
 
+class _AddRuntimeAssertsInInputConstraint(torch.fx.interpreter.Transformer):
```
Hmmph, I still am failing to get the big picture here.

Let's suppose that I trace and export a model that does if x.size(3) > 10 do something else do something else. Furthermore, let's suppose I didn't pass in any explicit constraints (which I assume means "try to make everything as dynamic as possible"). Then you will end up with a model which requires x.size(3) > 10 to be sound. But you aren't going to have a constraint for it (because it's dynamic by default). Do you still want to put the tests into the graph?

More generally, it seems to me that you are trying to solve the problem of "what to do with guards" by directly putting them in the graph. But does this make sense? Why is this not tested out of band?

tugsbayasgalan replied:
My thinking is that Avik's surfacing of guards will output those more fine-grained constraints directly to user (e.g x.size(3) > 10). And we suggest user to use those constraints directly on the input to the export call so that the exported module is more sound. And then, my pass will translate the new constraints into runtime assertions. 

One thing we can explore is, we can put the constraints derived from guards directly into graph after we finish exporting (not asking user to pass in the constraint).

ezyang replied:
I guess, my point is, it's inconsistent to only do directly given constraints, and not also do implicit constraints from guards.

tugsbayasgalan replied:
Yeah I agree. It should also convert the implicit constraints as well. Can do it in the follow up diff.

voznesenskym replied:
This is worth doing now, and is more important, and the primary reason for my requested changes.

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
+            dim = self.tracer.create_proxy(
+                "call_function", torch.ops.aten.sym_size, (arg, constraint.dim), {}
+            )
+            assert_msg = f"Input #{self.count}'s dimension #{constraint.dim} size is outside of supported dynamic range"
```
Nit, if you print orig_inp_id instead of self.count here, you can unconditionally increment self.count right after creating a new id without duplicating self.count = self.count + 1

tugsbayasgalan replied:
Hmm `orig_input_id` is actually a tensor python id tho (e.g id(tensor(5))).

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
+            dim = self.tracer.create_proxy(
+                "call_function", torch.ops.aten.sym_size, (arg, constraint.dim), {}
+            )
+            assert_msg = f"Input #{self.count}'s dimension #{constraint.dim} size is outside of supported dynamic range"
```
```suggestion
            assert_msg = f"Input #{self.count}'s dimension #{constraint.dim} size is outside of specified dynamic range [{min_int_val}, {max_int_val}]"
```

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
+
+    def call_function(self, target, args, kwargs):
+        arg = super().call_function(target, args, kwargs)
+        from torch._export.constraints import constrain_as_size
```
Consider moving all imports to top of the file, unless there is a circular dep issue?

tugsbayasgalan replied:
This was due to circular dep issue :(

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
+            min_val = kwargs["min"]
+            max_val = kwargs["max"]
+            gt = self.tracer.create_proxy(
+                "call_function", operator.gt, (args[0], min_val), {}
```
Should we use `operator.ge` instead? I think the ranges are inclusive on both ends. (Same changes needed for many other places in this PR)

tugsbayasgalan replied:
Yep good catch. It should be `ge` and `le`

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
+            )
+        return arg
+
+    def run_node(self, n):
```
What does this override cover? just `output` nodes?

tugsbayasgalan replied:
This is to populate `val` and `tensor_dict` in node.meta

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -643,6 +645,177 @@ def __bool__(self):
+        return r
+
+
+class _ChangeInputOutputSignature(torch.fx.interpreter.Transformer):
```
Some comments would be nice

tugsbayasgalan replied:
This was here before my change. I am not actually sure what it does...

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints into runtime assertions utilizing `aten._assert_async.msg` per dimension. 

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2.  Convert our inline constraints into runtime asserts

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Add runtime assert"


This PR introduces a new operator called `aten._assert_async.msg`, which allows passing a tensor value and assertion message as inputs. As part of TorchDynamo, we're replacing the use of torch._assert with this new operator so that `make_fx` also knows how to handle assertions.

Originally, we planned to create a dependency chain to introduce a fake control dependency, but this new implementation seems to work with AOTAutograd and friends, which will be demonstrated in the next pull request.

In addition, we also make input constraints and intermediate constraints into runtime assertions utilizing `aten._assert_async.msg`.

Future work:
1.  Assess whether we still need to introduce a fake control dependency
2. Explore adding non-async version of assert. 

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri reviewed (approved):
Looks good for a start. Follow-up work (among other things pointed out in the implementation itself) is to broaden the scope, by adding assertions for other things, like specializations, equalities, etc.

voznesenskym commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -405,6 +405,10 @@ void _assert_async_cpu(const Tensor& self) {
 }
 
+void _assert_async_msg_cpu(const Tensor& self, c10::string_view assert_msg) {
+  TORCH_CHECK(native::is_nonzero(self), assert_msg != "" ? assert_msg : "Assertion is failed");
```
Use AT_ASSERT

voznesenskym replied:
This message is sus because you might violate semantics preserving notions here. 

You also don't ever pass the tensor to the message. Which is fine, but kinda lame.

tugsbayasgalan replied:
Sorry can you elaborate more what do you mean?

voznesenskym commented on test/dynamo/test_repros.py:
```diff
@@ -2475,15 +2480,15 @@ def f(x):
         cnt = torch._dynamo.testing.CompileCounter()
 
-        opt_f = torch._dynamo.optimize(cnt, nopython=True)(f)
+        opt_f = torch._dynamo.optimize(cnt, nopython=True, dynamic=True)(f)
```
Redundant w/ patch.

voznesenskym reviewed (changes requested):
Needs more discussion.

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -642,6 +644,221 @@ def __bool__(self):
         )
 
 
+ConstraintSpec = namedtuple("ConstraintSpec", ["constraint_dim", "min_val", "max_val"])
```
No constrain spec notions in eval frame.

tugsbayasgalan replied:
Hmm the export API takes in Constraint notion as input tho

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -793,41 +1010,11 @@ def result_capturing_wrapper(*graph_inputs):
-            if "val" in self.current_node.meta:
-                r.node.meta["val"] = self.current_node.meta["val"]
-            return r
+    graph = _AddRuntimeAssertsInInputConstraint(
```
I don't understand why we want this as a default? Like, this feels like something that should exist entirely outside of dynamo - anyone is free to add any passes or transforms they want, but this is far too assumptive of a single usecase.

tugsbayasgalan replied:
We currently don't have agreed final export API yet. So there is no other place to put this pass at the moment. I can move it out of dynamo once we have that API. Are you ok with it if I hide it under a flag?

voznesenskym replied:
What I would like to discuss is the direction of passes like this in dynamo. To me, this feels like a laborious and assumptive default, or inclusion. If someone needs this, they can invoke a util we provide. 

w/r/t hiding it under a flag - export growing in flags and flag combinations is not a good direction. 

The burden of proof of including a new feature is that it is generally useable, and I think promotion of guards to asserts does not cross that bar. Offering a util that does it. and putting the burden on the .export caller to use it seems sufficient.

gmagogsfm replied:
Thanks for the comment @voznesenskym. Tugsuu and I analyzed the different options and we also feel that it is better to implement as a util that can be invoked after dynamo export call, so that dynamo can maintain the contract of "making assumptions, leave it to the caller as to how to enforce them"

voznesenskym commented:
I feel like a jerk rejecting this - because this is fundamentally a good change - I think we just need to think a little bit more about (1) where it should live and (2) What it should encompass (If we do this, I would really like to see this for all guards, not just the user directives)

CaoE commented:
@tugsbayasgalan May I know why not add cuda kernel for _assert_async.msg ?
